### PR TITLE
Add callback to manage fault-tolerance checkpoints

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/9130

This callback is not meant to be used publicly.
There are tests that cover this works.

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @rohitgr7 @ananthsub @daniellepintz @carmocca @ninginthecloud @akihironitta

### CODE REVIEW

File: pytorch_lightning/callbacks/fault_tolerance.py
Comment (carmocca) [inline pytorch_lightning/callbacks/fault_tolerance.py:1]: I tried to put this in `pytorch_lightning/utilities/auto_restart.py` initially but couldn't resolve the circular dependencies.

File: pytorch_lightning/trainer/trainer.py
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:764]: This was a tiny bug, but it's not necessary in the bug-fix because `model` can be used as a truthy check

**Comment by daniellepintz** [comment]
Hi @carmocca, any thoughts on https://github.com/PyTorchLightning/pytorch-lightning/issues/11768#issuecomment-1035469908?

File: pytorch_lightning/callbacks/fault_tolerance.py
Comment (ananthsub) [inline pytorch_lightning/callbacks/fault_tolerance.py:46]: what happens if there are exceptions only on some ranks? calling `save_checkpoint` afterward which relies on collectives will lead to a hang

Comment (carmocca) [reply pytorch_lightning/callbacks/fault_tolerance.py:46]: That's a great question. I guess there's nothing we can do in that case.
The best use case you can think for fault-tolerance is to prepare for your cloud machine  getting killed. Not so much for saving a broken run, especially if it already failed to communicate in a distributed setting.

This limitation applies to the previous code too.

File: CHANGELOG.md
Comment (daniellepintz) [inline CHANGELOG.md:36]:
```suggestion
- Added a private callback to manage the creation and deletion of fault-tolerance checkpoints ([#11862](https://github.com/PyTorchLightning/pytorch-lightning/pull/11862))
```

**Comment by daniellepintz** [comment]
@carmocca is this PR ready to go or is it waiting on something? I ask because I'd like to start working on https://github.com/PyTorchLightning/pytorch-lightning/issues/11768

**Comment by carmocca** [comment]
It's waiting on 3 approvals + CI working

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/callbacks/fault_tolerance.py
Comment (awaelchli) [inline pytorch_lightning/callbacks/fault_tolerance.py:49]: It's important that when we are in the exception handling, we don't call teardown on the hooks. Otherwise we would delete the file we just created. I'm wondering, should we add a test for this sequence to the hook tests in test_hooks.py?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/callbacks/fault_tolerance.py
Comment (carmocca) [reply pytorch_lightning/callbacks/fault_tolerance.py:49]: The auto restart tests would fail if this happened because the fault-tolerant checkpoint would be gone

File: pytorch_lightning/callbacks/fault_tolerance.py
Comment (tchaton) [reply pytorch_lightning/callbacks/fault_tolerance.py:1]: Can we make the file private, like pytorch_lightning/callbacks/_fault_tolerance.py

**Review by tchaton — APPROVED** [review]
LGTM !

**Pull request merged**
